src/data_chain/chain.rs

```diff
@@ -185,3 +171,309 @@ impl Chain {
+        // Relocated(B), Live(C).  Add the four of them in reverse order and assert after each is
+        // added that they're held in the `pending_blocks` until the final one is added and they can
+        // all be added to the chain in the order defined above.
+        let mut rng = SeededRng::thread_rng();
+        unwrap!(rust_sodium::init_with_rng(&mut rng));
+
+        let prefix = Prefix::default().pushed(rng.gen()).with_version(rng.gen());
+        let nodes = tests::create_full_infos(Some(prefix));
+
+        // Create Live(A) and Offline(A).
+        let node_a = FullInfo::within_range(5, &prefix.lower_bound(), &prefix.upper_bound());
+        let mut payload = NodeState {
+            state: State::ElderLive,
+            public_key: *node_a.public_info().sign_key(),
+            age: node_a.public_info().age(),
+            section: prefix,
+        };
+        let live_a = tests::create_block(payload, &nodes, false);
+        payload = NodeState {
+            state: State::ElderOffline,
+            public_key: *node_a.public_info().sign_key(),
+            age: node_a.public_info().age(),
+            section: prefix,
+        };
+        let offline_a = tests::create_block(payload, &nodes, false);
+
+        // Create Live(B) and Relocated(B).
+        let node_b = FullInfo::within_range(6, &prefix.lower_bound(), &prefix.upper_bound());
+        payload = NodeState {
+            state: State::ElderLive,
+            public_key: *node_b.public_info().sign_key(),
+            age: node_b.public_info().age(),
+            section: prefix,
+        };
+        let live_b = tests::create_block(payload, &nodes, false);
+        payload = NodeState {
+            state: State::ElderRelocated,
+            public_key: *node_b.public_info().sign_key(),
+            age: node_b.public_info().age(),
+            section: prefix,
+        };
+        let relocated_b = tests::create_block(payload, &nodes, false);
+
+        // Create Live(C).
+        let node_c = FullInfo::within_range(7, &prefix.lower_bound(), &prefix.upper_bound());
+        payload = NodeState {
+            state: State::ElderLive,
+            public_key: *node_c.public_info().sign_key(),
+            age: node_c.public_info().age(),
+            section: prefix,
+        };
+        let live_c = tests::create_block(payload, &nodes, false);
+
+        // Create chain containing just Live(A).
+        let mut chain = Chain::from_blocks(vec![live_a.clone()], vec![], nodes.len());
+        let mut expected_blocks = vec![live_a.clone()];
+        let mut expected_pendings = vec![];
+
+        // Add Live(C) - should be held in `pending_blocks` as there's no corresponding Offline/
+        // Demoted/Relocated in the `pending_blocks`.
+        chain.add_block(live_c.clone());
+        expected_pendings.push(live_c.clone());
+        assert_eq!(chain.blocks, expected_blocks);
+        assert_eq!(chain.pending_blocks, expected_pendings);
```

Comment:
I wonder whether it's a good idea to test private fields like this. My suggestion would be to ignore `pending_blocks` (it's an implementation detail, I don't think it should be tested) and only test the contents of `blocks`, but via a public accessor. Something like: `fn blocks(&self) -> &[Block<...>]`.

Reply:
I'm fine to put an accessor for `blocks`, but I think we'd want one for `pending_blocks` too.  So far we don't have a use for either, but I expect when we come to passing our chain to a joining node, we'll want those (rather than just serialising the full chain).  So, I'll add those two accessors and use them in the tests.